loadfile/loadfile.go

```diff
@@ -7,41 +7,168 @@ import (
+// directory path to be used in conjunction with files that do not
+// provide an absolute file path.
+type FileCache interface {
+	RootDir() string
+	LoadContext() error
+	GetByKey(fileKey string) (*ContextFile, bool)
+	AbsPathByKey(fileKey string) *string
+	ContentByKey(fileKey string) []byte
+	Contents() map[string][]byte
+	AbsPaths() map[string]string
+	Files() map[string]ContextFile
+}
+
+// NewFileCacheUsingContext creates a mapping of files to their absolute paths and
+// file contents. rootDir is the context directory (root directory) in
+// which the files can be found. 'files' is a mapping of the desired file
+// key to a relative or absolute file path.
+func NewFileCacheUsingContext(rootDir string, files map[string]string) (FileCache, error) {
 	absDir, err := filepath.Abs(rootDir)
 	if err != nil {
 		return nil, fmt.Errorf("error determining context directory absolute path %s (%w)", rootDir, err)
 	}
-	requiredFilesAbs := map[string]string{}
-	for key, f := range requiredFiles {
+	filesAbsPaths := map[string]ContextFile{}
+	for key, f := range files {
 		abspath := f
 		if !filepath.IsAbs(f) {
 			abspath = filepath.Join(absDir, f)
 		}
-		requiredFilesAbs[key] = abspath
+		filesAbsPaths[key] = ContextFile{
+			ID:           f,
+			AbsolutePath: abspath,
+		}
+	}
+	return &fileCache{
+		rootDir: absDir, files: filesAbsPaths}, nil
+}
+
+// NewFileCache creates a file cache where the fileContents is
+// a mapping of paths to file contents. The keys in fileContents
+// become the keys in Files, and the absolute paths and ID for
+// each context file.
+func NewFileCache(rootDir string, fileContents map[string][]byte) FileCache {
+	files := map[string]ContextFile{}
+	for key, content := range fileContents {
+		files[key] = ContextFile{
+			ID:           key,
+			AbsolutePath: key,
+			Content:      content,
+		}
+	}
+	return &fileCache{
+		rootDir: rootDir,
+		files:   files,
+	}
+}
+
+// LoadContext reads the content of each context file into Content.
+func (fc *fileCache) LoadContext() error {
+	result := map[string]ContextFile{}
+	var err error
+	for key, cf := range fc.files {
+		absPath := cf.AbsolutePath
+		fileData, err := os.ReadFile(filepath.Clean(absPath))
+		if err != nil {
+			return fmt.Errorf("error reading file %s (%w)", absPath, err)
+		}
+		result[key] = ContextFile{
+			ID:           cf.ID,
+			AbsolutePath: cf.AbsolutePath,
+			Content:      fileData,
+		}
+	}
+	fc.files = result
+	return err
+}
+
+// RootDir returns the root directory used by files with
+// relative file paths.
+func (fc *fileCache) RootDir() string {
+	return fc.rootDir
+}
+
+// GetByKey asks if the file cache contains the given fileKey.
+func (fc *fileCache) GetByKey(fileKey string) (*ContextFile, bool) {
+	cf, ok := fc.files[fileKey]
+	return &cf, ok
+}
+
+// AbsPathByKey returns the absolute file path of a given file key,
+// if it exists in the file cache, nil otherwise.
+func (fc *fileCache) AbsPathByKey(fileKey string) *string {
```

Comment:
Isn't it more conventional in Go to return an `err` value for errors?  The string pointer is cumbersome, and it's easy to return an empty string in the event of an error.

Reply:
I had that thought as well, but it seemed better to make it more like a map access. I don't usually like conflating `nil` with empty string. I can change the return value to `string`, so that it returns an empty string.

Reply:
Yes...my intention was that defining the function to return a string pointer solely so that we could return `nil` in the event of an error seemed anti-Go-ish.  If we define it to return an `err` value for errors, that covers the mechanism requirement, and then we can have it return a string instead of a pointer...we just have to figure out what value to return for the string value in the event of an error.  We presumably cannot return `nil`, but we can easily construct an empty string from a literal and return that.